src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSScoreUtils.java

```diff
@@ -0,0 +1,426 @@
+            }
+        } else if (unpairedHeader != null) {
+            header = unpairedHeader;
+        } else {
+            throw new UserException.BadInput("No headers were loaded");
+        }
+        return header;
+    }
+
+    /**
+     * Writes accessions contained in a SAM header that do not exist in the taxonomy database
+     */
+    public static void writeHeaderWarnings(final String path, final SAMFileHeader header, final PSTaxonomyDatabase taxDB,
+                                           final Logger logger) {
+        if (header != null && header.getSequenceDictionary() != null && header.getSequenceDictionary().getSequences() != null) {
+            final Set<String> unknownSequences = header.getSequenceDictionary().getSequences().stream()
+                    .map(SAMSequenceRecord::getSequenceName)
+                    .filter(name -> !taxDB.accessionToTaxId.containsKey(name))
+                    .collect(Collectors.toSet());
+            try {
+                final OutputStream file = BucketUtils.createFile(path);
+                for (final String acc : unknownSequences) {
+                    final String line = acc + "\n";
+                    file.write(line.getBytes(Charset.defaultCharset()));
+                }
+                file.close();
+            } catch (final IOException e) {
+                logger.warn("Could not write header warnings to " + path, e);
+            }
+        }
+
+    }
+
+    /**
+     * Gets taxonomic IDs of contigs that aligned sufficiently well to the reads. If both mates are present, returns the
+     * intersection of their hits. Also sets read tag HITS_TAG to a comma-separated list of the IDs.
+     */
+    static JavaRDD<Tuple2<Iterable<GATKRead>, PSPathogenAlignmentHit>> mapGroupedReadsToTax(final JavaRDD<Iterable<GATKRead>> pairs,
+                                                                                            final double minCoverage,
+                                                                                            final double minIdentity,
+                                                                                            final Broadcast<Map<String, String>> refNameToTaxBroadcast) {
+        return pairs.map(readIter -> {
+
+            final Collection<List<String>> taxIDLists = Utils.stream(readIter)
+                    //Flat map the taxID's of all alignments meeting the coverage/identity criteria.
+                    //Throw out duplicates so it returns a list of unique tax ID's for each read in the pair
+                    .flatMap(read -> getValidHits(read, minCoverage, minIdentity).stream()
+                            .map(contig -> refNameToTaxBroadcast.value().containsKey(contig) ? refNameToTaxBroadcast.value().get(contig) : null)
+                            .filter(Objects::nonNull).distinct()
+                    ).collect(Collectors.groupingBy(Function.identity())).values(); //Group the flattened stream by tax ID into lists
```

Comment:
Done. This whole section was a bit of a mess and did a lot of extra work for unpaired reads. It's cleaned up and a bit more readable now.